---
title: "Leaf shape microbiome"
author: "Julia Boyle"
date: "2023-07-31"
output: html_document
---
#Packages
```{r}
library(pander)
library(ape)
library(microbiomeSeq)
library(ranacapa)
library(adespatial)
library(plotly)
library(philr)
library(MicrobeR)
library(microbiome)
library(microbiomeutilities)
library(qiime2R)
library(phyloseq)
library(ANCOMBC)
library(plyr)
library(tidyverse)
library(car)
library(vegan)
library(RColorBrewer) 
library(randomcoloR)
library(cowplot)
library(lmerTest)
library(insight)
library(lme4)
library(compositions)
library(zCompositions)
library(phyloseqCompanion)
library(CoDaSeq)
library(microshades)#color blind friendly shades
library(speedyseq)
library(patchwork)
library(viridis)
library(picante)
library(mdthemes)
```

# Reading in data
```{r}
#Reading in metadata as it's own file
micrometa<- read.csv('ipomoea-taxa-metadata-allnumbers.csv',header = TRUE) #full metadata
micrometa$Shape<-factor(micrometa$Shape, levels=c("L", "M", "H"))
metadata<-read.table('ipomoea_metadata.tsv', header=TRUE, row.names = 1) #this only has plants that survived, so not all numbers between 1-240
metadata$Shape<-factor(metadata$Shape, levels=c("L", "M", "H"))
metadata$Line<-as.factor(metadata$Line)
metadata$Germination_cohort<-as.factor(metadata$Germination_cohort)
str(metadata)
metadata
featuretable<- read.csv('table-try3-10readsmin-filtered-cyano-mito.csv',row.names = 1, check.names = FALSE, header = TRUE) #this is the feature table as a .csv file
taxonomy<- read.csv('taxonomy_try3.csv') #the taxonomy file as a .csv file


phyglory <- qza_to_phyloseq("table-try3-10readsmin-filtered-cyano-mito-glory.qza", "rooted-tree-try3-glory.qza", "taxonomy-try3-glory.qza","ipomoea_metadata.tsv") 

sample_data(phyglory)$FloweringDOY<-as.factor(sample_data(phyglory)$FloweringDOY) 
sample_data(phyglory)$Days_to_flowering_factor<-as.factor(sample_data(phyglory)$Days_to_flowering)
sample_data(phyglory)$Line<-as.factor(sample_data(phyglory)$Line) 
sample_data(phyglory)$Shape<-factor(sample_data(phyglory)$Shape, levels=c("L", "M", "H"))
sample_data(phyglory)['sample_id'] <- row.names(sample_data(phyglory)) 
str(phyglory)
sample_data(phyglory)

# check for features of data  
summarize_phyloseq(phyglory)
print_ps(phyglory)
summary(sample_sums(phyglory))
ntaxa(phyglory)
nsamples(phyglory)
sample_names(phyglory)[1:5]  
rank_names(phyglory)  
sample_variables(phyglory)  
otu_table(phyglory)[1:5, 1:5]  
tax_table(phyglory)[1:10, 1:4]

#Checking taxa prevalence at phylum level
plot_taxa_prevalence(phyglory, "Phylum")
```

# Rarefaction of data
```{r}
# set seed
set.seed(1)
subsamples <- seq(0, 23100, by=100)[-1]
phygloryrarefactioncurve <- plot_alpha_rcurve(phyglory, index="observed", #phyglory is the file
                       subsamples = subsamples,
                       lower.conf = 0.5, 
                       upper.conf = 0.5,
                       group="Shape",
                       label.color = "brown3",
                       label.size = 3,
                       label.min = TRUE) 
print(phygloryrarefactioncurve) #M shape plateaus at a lower observed species number
```

```{r}
phyglory_rarefy <- rarefy_even_depth(phyglory, rngseed=1, sample.size=0.4*median(sample_sums(phyglory)), replace=F) # approx 1809 reads
summarize_phyloseq(phyglory_rarefy)
ntaxa(phyglory_rarefy)
nsamples(phyglory_rarefy)
```

# Centered log ratio transformation 
```{r}
#ASV feature table
featuretableCLR<-clr(featuretable+1) #add 1 to avoid taking log of 0, use compositions package
featuretableCLR<-as.data.frame(featuretableCLR)
featuretableCLR

#aggregated by Genus
phyglory_gen <- microbiome::aggregate_rare(phyglory, level = "Genus", detection = 0/100, prevalence = 30/100)#must be 0% relative abundance, and appear in at least 30% of samples
featuretableCLR_gen<-as.data.frame(otu_table(phyglory_gen))
#write.csv(featuretableCLR_gen, "phygloryCLRGenfeaturetable.csv")#transposed manually in excel, rows for Other and Unknown genera removed
featuretableCLRgen<-read.csv('phygloryCLRGenfeaturetable.csv',row.names = 1, check.names = FALSE, header = TRUE)
featuretableCLRgen<-clr(featuretableCLRgen+1) #add 1 to avoid taking log of 0
featuretableCLRgen<-as.data.frame(featuretableCLRgen)
featuretableCLRgen
```

# Leaf shape statistics
## Leaf area
```{r}
hist(micrometa$Leaf_area_cm2)
summary(lmer(Leaf_area_cm2~Shape+Germination_cohort + (1|Line) + (1|Block), data=micrometa))
Anova((lmer(Leaf_area_cm2~Shape+Germination_cohort + (1|Line) + (1|Block), data=micrometa)), type=3)

area<-metadata%>% 
ggplot(aes(x=Germination_cohort, y=Leaf_area_cm2, fill=Shape))+
  geom_boxplot()+ theme_classic() + ylab('Leaf area (cm^2)')+xlab('Germination cohort') +scale_fill_manual(labels = c("Lobe", "Heterozygote", "Whole"), values = c("#00CCCC", "#9999FF", "#FF3333"))
```
## Circularity
"4π×([Area]/([Perimeter]2)) with a value of 1.0 indicating a perfect circle. As the value approaches 0.0, it indicates an increasingly elongated shape."
```{r}
hist(micrometa$Circularity)
Anova((lmer(Circularity~Shape+Germination_cohort + (1|Line) + (1|Block), data=micrometa)), type=3)

circularity<-metadata%>% 
ggplot(aes(x=Shape, y=Circularity, fill=Shape))+
  geom_boxplot()+ theme_classic() + ylab('Circularity')+scale_fill_manual(labels = c("Lobe", "Heterozygote", "Whole"), values = c("#00CCCC", "#9999FF", "#FF3333")) + scale_x_discrete(limits=c("L","M","H"), labels=c("Lobe", "Heterozygote", "Whole"))+theme(legend.position = "none")

```
## Between cohorts
```{r}
chisq.test(x=micrometa$Germination_cohort, y=micrometa$Shape) #No significant differences in leaf genotypes between cohorts
chisq.test(x=micrometa$Block, y=micrometa$Shape) #no sig differences in leaf shape genotypes between blocks
```
## Hardy-Weinburg equilibrium test
```{r}
observed<-c(51, 112, 55) #51 lobe, 112 heterozygotes, 55 whole genotypes
expected<-c(54.5, 109, 54.5) #genotype frequencies if they were 1:2:1
chisq.test(observed, expected)
```

# Composition plots
Family level
```{r}
#making relative abundances objects for different taxonomic levels
phyglory_fam <- microbiome::aggregate_rare(phyglory, level = "Family", detection = 10/100, prevalence = 30/100)
phyglory.fam.rel <- microbiome::transform(phyglory_fam, "compositional")
phyglory.fam.rel <- phyglory %>%
  aggregate_rare(level = "Family", detection = 10/100, prevalence = 30/100) %>%
  microbiome::transform(transform = "compositional")

phyglory.fam.rel.germ1<-subset_samples(phyglory.fam.rel, Germination_cohort%in%c("First"))
phyglory.fam.rel.germ2<-subset_samples(phyglory.fam.rel, Germination_cohort%in%c("Second"))

palettegood <- distinctColorPalette(50)

leaf_germ1<-plot_composition(phyglory.fam.rel.germ1, sample.sort = "Sphingomonadaceae", x.label = "Shape") + theme_cowplot()+ theme(legend.position = "none") +scale_fill_manual(values =palettegood) + theme(axis.text.x = element_blank(), axis.text.y=element_text(size=10)) + ggtitle("First germination cohort") + theme(legend.title = element_text(size = 5), legend.text = element_text(face="italic"), axis.text = element_text(size=10) )+xlab("Samples")
leaf_germ1

leaf_germ2<-plot_composition(phyglory.fam.rel.germ2, sample.sort = "Sphingomonadaceae", x.label = "Shape") + theme_cowplot() +scale_fill_manual(values =palettegood) + theme(axis.text.x = element_blank(), axis.text.y=element_text(size=10)) + ggtitle("Second germination cohort") + theme(legend.title = element_text(size = 15),legend.text = element_text(face="italic"), axis.text = element_text(size=10) )+xlab("Samples")+ guides(fill=guide_legend(title="Bacterial Family"))
leaf_germ2

plot_composition(phyglory.fam.rel.germ1, sample.sort = "Shape", x.label = "Shape") + theme_cowplot()+ theme(legend.position = "none") +scale_fill_manual(values =palettegood) + theme( axis.text.y=element_text(size=10)) + ggtitle("First germination cohort") + theme(legend.title = element_text(size = 5), axis.text = element_text(size=10) )+xlab("Samples")
plot_composition(phyglory.fam.rel.germ2, sample.sort = "Shape", x.label = "Shape") + theme_cowplot()+ theme(legend.position = "none") +scale_fill_manual(values =palettegood) + theme( axis.text.y=element_text(size=10)) + ggtitle("First germination cohort") + theme(legend.title = element_text(size = 5), axis.text = element_text(size=10) )+xlab("Samples")
```

Get relative abundance of different families per sample for downstream purposes
```{r}
genera_relabund<-phyglory_rarefy %>% tax_glom(taxrank = "Genus") %>% 
        transform_sample_counts(function(x) {x/sum(x)}) %>% psmelt() %>% 
        dplyr::select(Genus, Sample, Abundance) %>% spread(Sample, Abundance)
tgenera_relabund = setNames(data.frame(t(genera_relabund[,-1])), genera_relabund[,1])
tgenera_relabund

rarefiedmeta<-as.data.frame(sample_data(phyglory_rarefy))
vector<-rarefiedmeta$sample_id #get order of phyloseq metadata samples

tgenera_relabund_sorted<-tgenera_relabund %>%
  slice(match(vector, rownames(tgenera_relabund))) #reorder the dataframe to be the same order as the phyloseq metadata

tgenera_relabund_sorted$Methylobacterium

sample_data(phyglory_rarefy)$Methylobacterium_relabundance<-tgenera_relabund_sorted$Methylobacterium
sample_data(phyglory_rarefy)$Sphingomonas_relabundance<-tgenera_relabund_sorted$Sphingomonas
sample_data(phyglory_rarefy)$Agrobacterium_relabundance<-tgenera_relabund_sorted$Agrobacterium
sample_data(phyglory_rarefy)$Deinococcus_relabundance<-tgenera_relabund_sorted$Deinococcus
sample_data(phyglory_rarefy)$Pseudomonas_relabundance<-tgenera_relabund_sorted$Pseudomonas
sample_data(phyglory_rarefy)$Arthrobacter_relabundance<-tgenera_relabund_sorted$Arthrobacter
```

# Core microbiome
```{r}
phyglory_core <- microbiome::aggregate_rare(phyglory_rarefy, level = "Genus", detection = 0/100, prevalence = 80/100)
core_members(phyglory_core) #determining taxa to highlight in graphs

phyglory_rarefy.germ1<-subset_samples(phyglory_rarefy, Germination_cohort%in%c("First"))
phyglory_rarefy.germ2<-subset_samples(phyglory_rarefy, Germination_cohort%in%c("Second"))
phyglory_core.germ1<- microbiome::aggregate_rare(phyglory_rarefy.germ1, level = "Genus", detection = 0/100, prevalence = 80/100)
phyglory_core.germ2<- microbiome::aggregate_rare(phyglory_rarefy.germ2, level = "Genus", detection = 0/100, prevalence = 80/100)
core_members(phyglory_core.germ1) #two microbial genera present only in germ cohort 1 at such high prevalence
core_members(phyglory_core.germ2)
```

# Alpha diversities
plot
```{r}
gloryrich<-estimate_richness(phyglory_rarefy)
gloryrich<-cbind(gloryrich, sample_data(phyglory_rarefy))
a.divglory<-gloryrich %>% 
  ggplot(aes(x=Germination_cohort, y=Observed, fill=Germination_cohort))+geom_boxplot()+theme_cowplot() +theme(axis.text.x = element_text(size=13), legend.position="none") +xlab("Germination Cohort") + scale_fill_manual("Germination_cohort", values=c("#FF9900", "#99CC33"))+ ylab("Observed richness")
a.divglory
```
statistical test
```{r}
gloryrichness<-estimate_richness(phyglory_rarefy) #calculates asv richness and shannon diversity
gloryrichness<-gloryrichness %>% mutate(evenness= Shannon/log(Observed)) #calculate evenness
gloryrichness
hist(gloryrichness$Shannon)
hist(log(gloryrichness$Observed))
hist(gloryrichness$evenness)

#shannon
Anova((lmer(Shannon~sample_data(phyglory_rarefy)$Shape+sample_data(phyglory_rarefy)$Germination_cohort + (1|sample_data(phyglory_rarefy)$Line) + (1|sample_data(phyglory_rarefy)$Block), data=gloryrichness)), type=3)
#Observed
Anova((lmer(log(Observed)~sample_data(phyglory_rarefy)$Shape+sample_data(phyglory_rarefy)$Germination_cohort + (1|sample_data(phyglory_rarefy)$Line) + (1|sample_data(phyglory_rarefy)$Block), data=gloryrichness)), type=3)
#Evenness
Anova((lmer(evenness~sample_data(phyglory_rarefy)$Shape+sample_data(phyglory_rarefy)$Germination_cohort + (1|sample_data(phyglory_rarefy)$Line) + (1|sample_data(phyglory_rarefy)$Block), data=gloryrichness)), type=3)
```

# Beta diversity
## PCoA
Leaf shape and germination cohort
```{r}
phyglory_rel_rare <- microbiome::transform(phyglory_rarefy, "compositional")
phyglory.ord.wuni_rare <- ordinate(phyglory_rel_rare, "PCoA", "unifrac", weighted=T)

b.div.wuniglory_rare <- plot_ordination(phyglory_rel_rare, phyglory.ord.wuni_rare, type= "samples", color= "Shape") + geom_point(size=3) + theme_cowplot() +ggtitle("")+ scale_color_manual("Leaf shape genotype",labels = c("Lobe", "Heterozygote", "Whole"), values = c("#00CCCC", "#9999FF", "#FF3333"))+ stat_ellipse()+xlab("Axis 1 [53.2%]")+ylab("Axis 2 [23.2%]")
print(b.div.wuniglory_rare)

b.div.wuniglory_rareG <- plot_ordination(phyglory_rel_rare, phyglory.ord.wuni_rare, type= "samples", color= "Germination_cohort") + geom_point(size=3) +ggtitle("")  + scale_color_manual("Germination cohort",values=c("#FF9900", "#99CC33")) +theme_cowplot() + stat_ellipse()+xlab("Axis 1 [53.2%]")+ylab("Axis 2 [23.2%]")
print(b.div.wuniglory_rareG)

print(plot_ordination(phyglory_rel_rare, phyglory.ord.wuni_rare, type= "samples", color= "Flowered_by_collection") + geom_point(size=3) +ggtitle("")  + scale_color_manual("Flowered",values=c("#99CC33", "#FF9900")) +theme_cowplot() + stat_ellipse()+xlab("Axis 1 [53.2%]")+ylab("Axis 2 [23.2%]"))
```

Core taxa driving the variation
```{r}
b.div.wuniglory_methylo <- plot_ordination(phyglory_rel_rare, phyglory.ord.wuni_rare, type= "samples", color= "Methylobacterium_relabundance") + geom_point(size=3) + ggtitle("Methylobacterium") +theme_cowplot()+xlab("Axis 1 [53.2%]")+ylab("Axis 2 [23.2%]")+scale_color_viridis(name="Relative abundance",discrete=FALSE, option="viridis")+theme(plot.title = element_text(face = "italic"))
print(b.div.wuniglory_methylo) #driving PC1

b.div.wuniglory_sphingo <- plot_ordination(phyglory_rel_rare, phyglory.ord.wuni_rare, type= "samples", color= "Sphingomonas_relabundance") + geom_point(size=3) + ggtitle("Sphingomonas") +theme_cowplot()+xlab("Axis 1 [53.2%]")+ylab("Axis 2 [23.2%]")+scale_color_viridis(name="Relative abundance",discrete=FALSE, option="viridis")+theme(plot.title = element_text(face = "italic"))
print(b.div.wuniglory_sphingo) #driving PC2

b.div.wuniglory_dein <- plot_ordination(phyglory_rel_rare, phyglory.ord.wuni_rare, type= "samples", color= "Deinococcus_relabundance") + geom_point(size=3) + ggtitle("Deinococcus") +theme_cowplot()+scale_color_viridis(name="Relative abundance",discrete=FALSE, option="viridis")+xlab("Axis 1 [53.2%]")+ylab("Axis 2 [23.2%]")+theme(plot.title = element_text(face = "italic"))
print(b.div.wuniglory_dein) #the couple dots out of place in methylobacterium plot

b.div.wuniglory_art <- plot_ordination(phyglory_rel_rare, phyglory.ord.wuni_rare, type= "samples", color= "Arthrobacter_relabundance") + geom_point(size=3) + ggtitle("Arthrobacter") +theme_cowplot()+scale_color_viridis(name="Relative abundance",discrete=FALSE, option="viridis")+xlab("Axis 1 [53.2%]")+ylab("Axis 2 [23.2%]")+theme(plot.title = element_text(face = "italic"))
print(b.div.wuniglory_art) #explains upper left corner
```

## Permanova and Permadisp
```{r}
set.seed(1)
wunifracdist<-distance(phyglory_rel_rare, method="wunifrac", type="samples")
wunifracdist

metaglory_rare<-as.data.frame(as.matrix(sample_data(phyglory_rel_rare)))

adonis2(wunifracdist ~ Shape+Germination_cohort, data = metaglory_rare, permutations=999) #sig germination cohort
anova(betadisper(d=wunifracdist, group=metaglory_rare$Germination_cohort)) #sig germination cohort. This means the sig germ cohort wunifrac distances could be due to dispersion differences. Addressed below
```

Checking that having slightly uneven sample sizes in germ 1 and 2 does not affect the significance of our permanova and permadisp
```{r}
phyglory_rel_rareG1<-subset_samples(phyglory_rel_rare, Germination_cohort%in%c("First"))
phyglory_rel_rareG2<-subset_samples(phyglory_rel_rare, Germination_cohort%in%c("Second"))
phyglory_rel_rareG1 #77 samples
phyglory_rel_rareG2 #106 samples

sample_ps <- function(phyglory_rel_rareG2, FUN = sample, ...){ 
  ids <- sample_names(phyglory_rel_rareG2)
  sampled_ids <- FUN(ids, ...) #randomly sample from id sample numbers
  ps <- prune_samples(sampled_ids, phyglory_rel_rareG2) #prune germ2 to the randomized set
  ps_joined<-merge_phyloseq(ps, phyglory_rel_rareG1) #merge together germ1 with the reshuffled germ2
  return(ps_joined)
}
set.seed(1)
permanovalist <- list() #make empty lists for the test results
permadisplist_germ<-list()
permadisplist_shape<-list()
for (i in 1:100){ #try with reshuffling germ2 data 100 times
  ps_joined<-sample_ps(phyglory_rel_rareG2, size=77, replace=FALSE) #do the random shuffling so that germ2 only has 77 samples
  wunifracdistjoined<-distance(ps_joined, method="wunifrac", type="samples") #create the distance matrix
  permanovalist[[i]]<-(adonis2(wunifracdistjoined ~ Shape+Germination_cohort, data = as.data.frame(as.matrix(sample_data(ps_joined))), permutations=999))
  permadisplist_germ[[i]]<-(anova(betadisper(d=wunifracdistjoined, group=as.data.frame(as.matrix(sample_data(ps_joined)))$Germination_cohort)))
  permadisplist_shape[[i]]<-(anova(betadisper(d=wunifracdistjoined, group=as.data.frame(as.matrix(sample_data(ps_joined)))$Shape)))
}

#extract F and p from the permanova
permanova<-data.frame(matrix(unlist(permanovalist), nrow=length(permanovalist), byrow=TRUE))
permanova<- dplyr::select(permanova, X13, X14, X17, X18)
colnames(permanova) <- c("Shape_F", "Germ_F", "Shape_p", "Germ_p")
permanova

#extract F and p for the permadisp of germination
permadisp_germ<-data.frame(matrix(unlist(permadisplist_germ), nrow=length(permadisplist_germ), byrow=TRUE))
permadisp_germ<-dplyr::select(permadisp_germ, X7,X9)
colnames(permadisp_germ) <- c("Germ_F", "Germ_p" )
permadisp_germ

hist(permanova$Shape_p, breaks=12)
hist1<-hist(permanova$Germ_p)
hist2<-hist(permadisp_germ$Germ_p)
plot(hist1, xlab=("Germination cohort p value"), main=("PERMANOVA"))
plot(hist2, xlab=("Germination cohort p value"), main=("PERMADISP"))
```

# Network analysis
```{r}
library(SpiecEasi)
library(igraph)
library(MCL)
library(qgraph)
```

Aggregated at genus
```{r}
phyglory_rarefy

glory_c1<-subset_samples(phyglory_rarefy, Germination_cohort%in%c("First"))
glory_c2<-subset_samples(phyglory_rarefy, Germination_cohort%in%c("Second"))
glory_L<-subset_samples(phyglory_rarefy, Shape%in%c("L"))
glory_M<-subset_samples(phyglory_rarefy, Shape%in%c("M"))
glory_H<-subset_samples(phyglory_rarefy, Shape%in%c("H"))


glory_c1<-t(abundances(aggregate_taxa(glory_c1, "Genus")))
glory_c2<-t(abundances(aggregate_taxa(glory_c2, "Genus")))
glory_L<-t(abundances(aggregate_taxa(glory_L, "Genus")))
glory_M<-t(abundances(aggregate_taxa(glory_M, "Genus")))
glory_H<-t(abundances(aggregate_taxa(glory_H, "Genus")))

glory_c1<-glory_c1[, colnames(glory_c1) != "Unknown"] #remove unknown columns
glory_c2<-glory_c2[, colnames(glory_c2) != "Unknown"]
glory_L<-glory_L[, colnames(glory_L) != "Unknown"]
glory_M<-glory_M[, colnames(glory_M) != "Unknown"]
glory_H<-glory_H[, colnames(glory_H) != "Unknown"]

glory_c1<-glory_c1[, which(colSums(glory_c1) != 0)] #remove columns with all 0s
glory_c2<-glory_c2[, which(colSums(glory_c2) != 0)]
glory_L<-glory_L[, which(colSums(glory_L) != 0)]
glory_M<-glory_M[, which(colSums(glory_M) != 0)]
glory_H<-glory_H[, which(colSums(glory_H) != 0)]

glory_c1Genusnames<-colnames(glory_c1)
glory_c2Genusnames<-colnames(glory_c2)
glory_LGenusnames<-colnames(glory_L)
glory_MGenusnames<-colnames(glory_M)
glory_HGenusnames<-colnames(glory_H)
```
## Lobe LL
Calculating Sparcc Pvalues
```{r}
set.seed(1)
#https://biovcnet.github.io/_pages/NetworkScience_SparCC.nb.html
#Calculating Sparcc Pvalues
sparcc.matrix_glory_L<-sparcc(glory_L)
sparcc.matrix_glory_L.permuted <- sparccboot(glory_L, R = 100)
glory_L.permuted.pvals <- pval.sparccboot(sparcc.matrix_glory_L.permuted)
data.frame(glory_L.permuted.pvals$cors, glory_L.permuted.pvals$pvals) %>% head
cors <- glory_L.permuted.pvals$cors
pvals <- glory_L.permuted.pvals$pvals
sparCCpcors_glory_L <- diag(0.5, nrow = dim(sparcc.matrix_glory_L$Cor)[1], ncol = dim(sparcc.matrix_glory_L$Cor)[1])
sparCCpcors_glory_L[upper.tri(sparCCpcors_glory_L, diag=FALSE)] <- cors
sparCCpcors_glory_L <- sparCCpcors_glory_L + t(sparCCpcors_glory_L)

sparCCpval_glory_L <- diag(0.5, nrow = dim(sparcc.matrix_glory_L$Cor)[1], ncol = dim(sparcc.matrix_glory_L$Cor)[1])
sparCCpval_glory_L[upper.tri(sparCCpval_glory_L, diag=FALSE)] <- pvals
sparCCpval_glory_L<- sparCCpval_glory_L + t(sparCCpval_glory_L)

rownames(sparCCpcors_glory_L) <- colnames(glory_L)
colnames(sparCCpcors_glory_L) <- colnames(glory_L)
rownames(sparCCpval_glory_L) <- colnames(glory_L)
colnames(sparCCpval_glory_L) <- colnames(glory_L)

sparCCpcors_glory_L[1:5, 1:5]
sparCCpval_glory_L[1:5, 1:5]
```
```{r}
# Get lower triangle of the correlation matrix
  get_lower_tri<-function(cormat){
    cormat[upper.tri(cormat)] <- NA
    return(cormat)
  }
  # Get upper triangle of the correlation matrix
  get_upper_tri <- function(cormat){
    cormat[lower.tri(cormat)]<- NA
    return(cormat)
  }

reorder_cormat <- function(cormat){
# Use correlation between variables as distance
dd <- as.dist((1-cormat)/2)
hc <- hclust(dd)
cormat <-cormat[hc$order, hc$order]
}
reorder_cor_and_p <- function(cormat, pmat){
  dd <- as.dist((1-cormat)/2)
  hc <- hclust(dd)
  cormat <-cormat[hc$order, hc$order]
  pmat <- pmat[hc$order, hc$order]
  list(r = cormat, p = pmat)
}
reordered_all_sparcc_glory_L <- reorder_cor_and_p(sparCCpcors_glory_L, sparCCpval_glory_L)
reordered_sparccCor_glory_L <- reordered_all_sparcc_glory_L$r
reordered_sparccP_glory_L<- reordered_all_sparcc_glory_L$p


sparccCor_processed_glory_L <- reordered_sparccCor_glory_L  %>% get_upper_tri() %>% reshape2::melt() %>% na.omit() %>% dplyr::rename(cor = value)
sparccP_processed_glory_L <- reordered_sparccP_glory_L  %>% get_upper_tri() %>% reshape2::melt() %>% na.omit() %>% dplyr::rename(p = value)

# join the two data frames
SparccP_glory_L <- left_join(sparccCor_processed_glory_L, sparccP_processed_glory_L, by = c("Var1", "Var2")) %>%
  # # remove self correlations
  # filter(Var1 != Var2) %>% 
  # calculate the false discovery rate to adjust for multiple p values
  mutate(fdr = p.adjust(p, method = "fdr"))
# Correlation matrix with p values
sparccOkP_glory_L <- SparccP_glory_L%>% filter(fdr < 0.05) 
sparccOkP_glory_L
sparccOkP_glory_L %>% filter(cor>0)
sparccOkP_glory_L %>% filter(cor<0)
```
Plotting only significant correlations
```{r}
glory_L_sig.net<-graph_from_data_frame(sparccOkP_glory_L[,1:3], directed=F, vertices=NULL)
plot(glory_L_sig.net)
```
Hub detection
```{r}
#Hub detection 
net_glory_L<-glory_L_sig.net
net.cn_glory_L<-closeness(net_glory_L) #measures centrality
net.bn_glory_L<-betweenness(net_glory_L) #measures centrality
net.pr_glory_L<-page_rank(net_glory_L)$vector 
net.hs_glory_L<-hub_score(net_glory_L)$vector
#Sort the species based on hubbiness score 
net.hs.sort_glory_L<-sort(net.hs_glory_L,decreasing=TRUE) 
#Choose the top 5 keystone species 
net.hs.top5_glory_L<-head(net.hs.sort_glory_L,n=5)
print(net.hs.top5_glory_L)
```
Cluster detection
```{r}
#Get clusters 
wt_glory_L<-walktrap.community(net_glory_L) 
ml_glory_L<-multilevel.community(net_glory_L) 
#Get membership of walktrap clusters 
membership(wt_glory_L)
wt_glory_Lmembership<-membership(wt_glory_L)
wt_glory_Lmembership<-as.data.frame(as.matrix(wt_glory_Lmembership))
colnames(wt_glory_Lmembership)<-'WT_cluster'
#Compare clusters detected by different methods 
compare(membership(wt_glory_L),membership(ml_glory_L))
#Plot clusters as dendrogram 
plot_dendrogram(wt_glory_L)
```
Modularity and articulation points
```{r}
#Calculate modularity 
modularity(net_glory_L,membership(wt_glory_L))
#Find articulation points 
AP_glory_L<-articulation.points(net_glory_L)#(Nodes so central that their removal will break the network into more components)
```
Network features
```{r}
nodes_glory_L<-V(net_glory_L) #vectors of nodes 
edges_glory_L<-V(net_glory_L) 
node.names_glory_L<-V(net_glory_L)$name 
num.nodes_glory_L<-vcount(net_glory_L) #number of nodes
print(num.nodes_glory_L) #(number of Asvs)
num.edges_glory_L<-ecount(net_glory_L) #number of edges (number of lines between nodes)
print(num.edges_glory_L)
clustering_coeff_glory_L<-transitivity(net_glory_L,type="global") #transensitivity
print(clustering_coeff_glory_L)
```
Plot
```{r}
plot(wt_glory_L,net_glory_L)
plot(wt_glory_L,net_glory_L, layout=layout.circle)
plot(wt_glory_L,net_glory_L, layout=layout.reingold.tilford)
plot(wt_glory_L,net_glory_L, vertex.label.cex = 0.4, frame.color=NA, main="Lobe LL") #save 8x8
```
## Het Ll
Calculating Sparcc Pvalues
```{r}
set.seed(1)
#https://biovcnet.github.io/_pages/NetworkScience_SparCC.nb.html
#Calculating Sparcc Pvalues
sparcc.matrix_glory_M<-sparcc(glory_M)
sparcc.matrix_glory_M.permuted <- sparccboot(glory_M, R = 100)
glory_M.permuted.pvals <- pval.sparccboot(sparcc.matrix_glory_M.permuted)
data.frame(glory_M.permuted.pvals$cors, glory_M.permuted.pvals$pvals) %>% head
cors <- glory_M.permuted.pvals$cors
pvals <- glory_M.permuted.pvals$pvals
sparCCpcors_glory_M <- diag(0.5, nrow = dim(sparcc.matrix_glory_M$Cor)[1], ncol = dim(sparcc.matrix_glory_M$Cor)[1])
sparCCpcors_glory_M[upper.tri(sparCCpcors_glory_M, diag=FALSE)] <- cors
sparCCpcors_glory_M <- sparCCpcors_glory_M + t(sparCCpcors_glory_M)

sparCCpval_glory_M <- diag(0.5, nrow = dim(sparcc.matrix_glory_M$Cor)[1], ncol = dim(sparcc.matrix_glory_M$Cor)[1])
sparCCpval_glory_M[upper.tri(sparCCpval_glory_M, diag=FALSE)] <- pvals
sparCCpval_glory_M<- sparCCpval_glory_M + t(sparCCpval_glory_M)

rownames(sparCCpcors_glory_M) <- colnames(glory_M)
colnames(sparCCpcors_glory_M) <- colnames(glory_M)
rownames(sparCCpval_glory_M) <- colnames(glory_M)
colnames(sparCCpval_glory_M) <- colnames(glory_M)

sparCCpcors_glory_M[1:5, 1:5]
sparCCpval_glory_M[1:5, 1:5]
```
```{r}
reordered_all_sparcc_glory_M <- reorder_cor_and_p(sparCCpcors_glory_M, sparCCpval_glory_M)
reordered_sparccCor_glory_M <- reordered_all_sparcc_glory_M$r
reordered_sparccP_glory_M<- reordered_all_sparcc_glory_M$p

sparccCor_processed_glory_M <- reordered_sparccCor_glory_M  %>% get_upper_tri() %>% reshape2::melt() %>% na.omit() %>% dplyr::rename(cor = value)
sparccP_processed_glory_M <- reordered_sparccP_glory_M  %>% get_upper_tri() %>% reshape2::melt() %>% na.omit() %>% dplyr::rename(p = value)

# join the two data frames
SparccP_glory_M <- left_join(sparccCor_processed_glory_M, sparccP_processed_glory_M, by = c("Var1", "Var2")) %>%
  # # remove self correlations
  # filter(Var1 != Var2) %>% 
  # calculate the false discovery rate to adjust for multiple p values
  mutate(fdr = p.adjust(p, method = "fdr"))
# Correlation matrix with p values
sparccOkP_glory_M <- SparccP_glory_M%>% filter(fdr < 0.05) 
sparccOkP_glory_M
sparccOkP_glory_M %>% filter(cor>0)
sparccOkP_glory_M %>% filter(cor<0)
```
Plotting only significant correlations
```{r}
glory_M_sig.net<-graph_from_data_frame(sparccOkP_glory_M[,1:3], directed=F, vertices=NULL)
plot(glory_M_sig.net)
```
Hub detection
```{r}
#Hub detection 
net_glory_M<-glory_M_sig.net
net.cn_glory_M<-closeness(net_glory_M) #measures centrality
print(net.cn_glory_M)#gives a measure of centrality for each ASV
net.bn_glory_M<-betweenness(net_glory_M) #measures centrality
net.pr_glory_M<-page_rank(net_glory_M)$vector 
net.hs_glory_M<-hub_score(net_glory_M)$vector
#Sort the species based on hubbiness score 
net.hs.sort_glory_M<-sort(net.hs_glory_M,decreasing=TRUE) 
#Choose the top 5 keystone species 
net.hs.top5_glory_M<-head(net.hs.sort_glory_M,n=5)
print(net.hs.top5_glory_M)
```
Cluster detection
```{r}
#Get clusters 
wt_glory_M<-walktrap.community(net_glory_M) 
ml_glory_M<-multilevel.community(net_glory_M) 
#Get membership of walktrap clusters 
membership(wt_glory_M)
wt_glory_Mmembership<-membership(wt_glory_M)
wt_glory_Mmembership<-as.data.frame(as.matrix(wt_glory_Mmembership))
colnames(wt_glory_Mmembership)<-'WT_cluster'
#Compare clusters detected by different methods 
compare(membership(wt_glory_M),membership(ml_glory_M))
#Plot clusters as dendrogram 
plot_dendrogram(wt_glory_M)
```
Modularity and articulation points
```{r}
#Calculate modularity 
modularity(net_glory_M,membership(wt_glory_M))
#Find articulation points 
AP_glory_M<-articulation.points(net_glory_M)#(Nodes so central that their removal will break the network into more components)
```
Network features
```{r}
nodes_glory_M<-V(net_glory_M) #vectors of nodes 
edges_glory_M<-V(net_glory_M) 
node.names_glory_M<-V(net_glory_M)$name 
num.nodes_glory_M<-vcount(net_glory_M) #number of nodes
print(num.nodes_glory_M) #(number of Asvs)
num.edges_glory_M<-ecount(net_glory_M) #number of edges (number of lines between nodes)
print(num.edges_glory_M)
clustering_coeff_glory_M<-transitivity(net_glory_M,type="global") #transensitivity
print(clustering_coeff_glory_M)
```
Plot
```{r}
plot(wt_glory_M,net_glory_M) 
plot(wt_glory_M,net_glory_M, vertex.label.cex = 0.4, frame.color=NA, main="Heterozygote Ll")
```
## Heart ll
Calculating Sparcc Pvalues
```{r}
set.seed(1)
#https://biovcnet.github.io/_pages/NetworkScience_SparCC.nb.html
#Calculating Sparcc Pvalues
sparcc.matrix_glory_H<-sparcc(glory_H)
sparcc.matrix_glory_H.permuted <- sparccboot(glory_H, R = 100)
glory_H.permuted.pvals <- pval.sparccboot(sparcc.matrix_glory_H.permuted)
data.frame(glory_H.permuted.pvals$cors, glory_H.permuted.pvals$pvals) %>% head
cors <- glory_H.permuted.pvals$cors
pvals <- glory_H.permuted.pvals$pvals
sparCCpcors_glory_H <- diag(0.5, nrow = dim(sparcc.matrix_glory_H$Cor)[1], ncol = dim(sparcc.matrix_glory_H$Cor)[1])
sparCCpcors_glory_H[upper.tri(sparCCpcors_glory_H, diag=FALSE)] <- cors
sparCCpcors_glory_H <- sparCCpcors_glory_H + t(sparCCpcors_glory_H)

sparCCpval_glory_H <- diag(0.5, nrow = dim(sparcc.matrix_glory_H$Cor)[1], ncol = dim(sparcc.matrix_glory_H$Cor)[1])
sparCCpval_glory_H[upper.tri(sparCCpval_glory_H, diag=FALSE)] <- pvals
sparCCpval_glory_H<- sparCCpval_glory_H + t(sparCCpval_glory_H)

rownames(sparCCpcors_glory_H) <- colnames(glory_H)
colnames(sparCCpcors_glory_H) <- colnames(glory_H)
rownames(sparCCpval_glory_H) <- colnames(glory_H)
colnames(sparCCpval_glory_H) <- colnames(glory_H)

sparCCpcors_glory_H[1:5, 1:5]
sparCCpval_glory_H[1:5, 1:5]
```
```{r}
reordered_all_sparcc_glory_H <- reorder_cor_and_p(sparCCpcors_glory_H, sparCCpval_glory_H)
reordered_sparccCor_glory_H <- reordered_all_sparcc_glory_H$r
reordered_sparccP_glory_H<- reordered_all_sparcc_glory_H$p

sparccCor_processed_glory_H <- reordered_sparccCor_glory_H  %>% get_upper_tri() %>% reshape2::melt() %>% na.omit() %>% dplyr::rename(cor = value)
sparccP_processed_glory_H <- reordered_sparccP_glory_H  %>% get_upper_tri() %>% reshape2::melt() %>% na.omit() %>% dplyr::rename(p = value)

# join the two data frames
SparccP_glory_H <- left_join(sparccCor_processed_glory_H, sparccP_processed_glory_H, by = c("Var1", "Var2")) %>%
  # # remove self correlations
  # filter(Var1 != Var2) %>% 
  # calculate the false discovery rate to adjust for multiple p values
  mutate(fdr = p.adjust(p, method = "fdr"))
# Correlation matrix with p values
sparccOkP_glory_H <- SparccP_glory_H%>% filter(fdr < 0.05) 
sparccOkP_glory_H
sparccOkP_glory_H %>% filter(cor>0)
sparccOkP_glory_H %>% filter(cor<0)
```
Plotting only significant correlations
```{r}
glory_H_sig.net<-graph_from_data_frame(sparccOkP_glory_H[,1:3], directed=F, vertices=NULL)
plot(glory_H_sig.net)
```
Hub detection
```{r}
#Hub detection 
net_glory_H<-glory_H_sig.net
net.cn_glory_H<-closeness(net_glory_H) #measures centrality
net.bn_glory_H<-betweenness(net_glory_H) #measures centrality
net.pr_glory_H<-page_rank(net_glory_H)$vector 
net.hs_glory_H<-hub_score(net_glory_H)$vector
#Sort the species based on hubbiness score 
net.hs.sort_glory_H<-sort(net.hs_glory_H,decreasing=TRUE) 
#Choose the top 5 keystone species 
net.hs.top5_glory_H<-head(net.hs.sort_glory_H,n=5)
print(net.hs.top5_glory_H)
```
Cluster detection
```{r}
#Get clusters 
wt_glory_H<-walktrap.community(net_glory_H) 
ml_glory_H<-multilevel.community(net_glory_H) 
#Get membership of walktrap clusters 
membership(wt_glory_H)
wt_glory_Hmembership<-membership(wt_glory_H)
wt_glory_Hmembership<-as.data.frame(as.matrix(wt_glory_Hmembership))
colnames(wt_glory_Hmembership)<-'WT_cluster'
#Compare clusters detected by different methods 
compare(membership(wt_glory_H),membership(ml_glory_H))
#Plot clusters as dendrogram 
plot_dendrogram(wt_glory_H)
```
Modularity and articulation points
```{r}
#Calculate modularity 
modularity(net_glory_H,membership(wt_glory_H))
#Find articulation points 
AP_glory_H<-articulation.points(net_glory_H)#(Nodes so central that their removal will break the network into more components)
```
Network features
```{r}
nodes_glory_H<-V(net_glory_H) #vectors of nodes 
edges_glory_H<-V(net_glory_H) 
node.names_glory_H<-V(net_glory_H)$name 
num.nodes_glory_H<-vcount(net_glory_H) #number of nodes
print(num.nodes_glory_H) #(number of Asvs)
num.edges_glory_H<-ecount(net_glory_H) #number of edges (number of lines between nodes)
print(num.edges_glory_H)
clustering_coeff_glory_H<-transitivity(net_glory_H,type="global") #transensitivity
print(clustering_coeff_glory_H)
```
Plot
```{r}
plot(wt_glory_H,net_glory_H) 
plot(wt_glory_H,net_glory_H, vertex.label.cex = 0.4, frame.color=NA, main="Whole ll")
```

## Germ cohort 1
Calculating Sparcc Pvalues
```{r}
set.seed(1)
#https://biovcnet.github.io/_pages/NetworkScience_SparCC.nb.html
#Calculating Sparcc Pvalues
sparcc.matrix_glory_c1<-sparcc(glory_c1)
sparcc.matrix_glory_c1.permuted <- sparccboot(glory_c1, R = 100)
glory_c1.permuted.pvals <- pval.sparccboot(sparcc.matrix_glory_c1.permuted)
data.frame(glory_c1.permuted.pvals$cors, glory_c1.permuted.pvals$pvals) %>% head
cors <- glory_c1.permuted.pvals$cors
pvals <- glory_c1.permuted.pvals$pvals
sparCCpcors_glory_c1 <- diag(0.5, nrow = dim(sparcc.matrix_glory_c1$Cor)[1], ncol = dim(sparcc.matrix_glory_c1$Cor)[1])
sparCCpcors_glory_c1[upper.tri(sparCCpcors_glory_c1, diag=FALSE)] <- cors
sparCCpcors_glory_c1 <- sparCCpcors_glory_c1 + t(sparCCpcors_glory_c1)

sparCCpval_glory_c1 <- diag(0.5, nrow = dim(sparcc.matrix_glory_c1$Cor)[1], ncol = dim(sparcc.matrix_glory_c1$Cor)[1])
sparCCpval_glory_c1[upper.tri(sparCCpval_glory_c1, diag=FALSE)] <- pvals
sparCCpval_glory_c1<- sparCCpval_glory_c1 + t(sparCCpval_glory_c1)

rownames(sparCCpcors_glory_c1) <- colnames(glory_c1)
colnames(sparCCpcors_glory_c1) <- colnames(glory_c1)
rownames(sparCCpval_glory_c1) <- colnames(glory_c1)
colnames(sparCCpval_glory_c1) <- colnames(glory_c1)

sparCCpcors_glory_c1[1:5, 1:5]
sparCCpval_glory_c1[1:5, 1:5]
```
```{r}
reordered_all_sparcc_glory_c1 <- reorder_cor_and_p(sparCCpcors_glory_c1, sparCCpval_glory_c1)
reordered_sparccCor_glory_c1 <- reordered_all_sparcc_glory_c1$r
reordered_sparccP_glory_c1<- reordered_all_sparcc_glory_c1$p

sparccCor_processed_glory_c1 <- reordered_sparccCor_glory_c1  %>% get_upper_tri() %>% reshape2::melt() %>% na.omit() %>% dplyr::rename(cor = value)
sparccP_processed_glory_c1 <- reordered_sparccP_glory_c1  %>% get_upper_tri() %>% reshape2::melt() %>% na.omit() %>% dplyr::rename(p = value)

# join the two data frames
SparccP_glory_c1 <- left_join(sparccCor_processed_glory_c1, sparccP_processed_glory_c1, by = c("Var1", "Var2")) %>%
  # # remove self correlations
  # filter(Var1 != Var2) %>% 
  # calculate the false discovery rate to adjust for multiple p values
  mutate(fdr = p.adjust(p, method = "fdr"))
# Correlation matrix with p values
sparccOkP_glory_c1 <- SparccP_glory_c1%>% filter(fdr < 0.05) 
sparccOkP_glory_c1
sparccOkP_glory_c1 %>% filter(cor>0)
sparccOkP_glory_c1 %>% filter(cor<0)
```
Plotting only significant correlations
```{r}
glory_c1_sig.net<-graph_from_data_frame(sparccOkP_glory_c1[,1:3], directed=F, vertices=NULL)
plot(glory_c1_sig.net)
```
Hub detection
```{r}
#Hub detection 
net_glory_c1<-glory_c1_sig.net
net.cn_glory_c1<-closeness(net_glory_c1) #measures centrality
net.bn_glory_c1<-betweenness(net_glory_c1) #measures centrality
net.pr_glory_c1<-page_rank(net_glory_c1)$vector 
net.hs_glory_c1<-hub_score(net_glory_c1)$vector
#Sort the species based on hubbiness score 
net.hs.sort_glory_c1<-sort(net.hs_glory_c1,decreasing=TRUE) 
#Choose the top 5 keystone species 
net.hs.top5_glory_c1<-head(net.hs.sort_glory_c1,n=5)
print(net.hs.top5_glory_c1)
```
Cluster detection
```{r}
#Get clusters 
wt_glory_c1<-walktrap.community(net_glory_c1) 
ml_glory_c1<-multilevel.community(net_glory_c1) 
#Get membership of walktrap clusters 
membership(wt_glory_c1)
wt_glory_c1membership<-membership(wt_glory_c1)
wt_glory_c1membership<-as.data.frame(as.matrix(wt_glory_c1membership))
colnames(wt_glory_c1membership)<-'WT_cluster'
#Compare clusters detected by different methods 
compare(membership(wt_glory_c1),membership(ml_glory_c1))
#Plot clusters as dendrogram 
plot_dendrogram(wt_glory_c1)
```
Modularity and articulation points
```{r}
#Calculate modularity 
modularity(net_glory_c1,membership(wt_glory_c1))
#Find articulation points 
AP_glory_c1<-articulation.points(net_glory_c1)#(Nodes so central that their removal will break the network into more components)
```
Network features
```{r}
nodes_glory_c1<-V(net_glory_c1) #vectors of nodes 
edges_glory_c1<-V(net_glory_c1) 
node.names_glory_c1<-V(net_glory_c1)$name 
num.nodes_glory_c1<-vcount(net_glory_c1) #number of nodes
print(num.nodes_glory_c1) #(number of Asvs)
num.edges_glory_c1<-ecount(net_glory_c1) #number of edges (number of lines between nodes)
print(num.edges_glory_c1)
clustering_coeff_glory_c1<-transitivity(net_glory_c1,type="global") #transensitivity
print(clustering_coeff_glory_c1)
```
Plot
```{r}
plot(wt_glory_c1,net_glory_c1) 
plot(wt_glory_c1,net_glory_c1, vertex.label.cex = 0.4, frame.color=NA, main="Older germination cohort")
```

## Germ cohort 2
Calculating Sparcc Pvalues
```{r}
set.seed(1)
#https://biovcnet.github.io/_pages/NetworkScience_SparCC.nb.html
#Calculating Sparcc Pvalues
sparcc.matrix_glory_c2<-sparcc(glory_c2)
sparcc.matrix_glory_c2.permuted <- sparccboot(glory_c2, R = 100)
glory_c2.permuted.pvals <- pval.sparccboot(sparcc.matrix_glory_c2.permuted)
data.frame(glory_c2.permuted.pvals$cors, glory_c2.permuted.pvals$pvals) %>% head
cors <- glory_c2.permuted.pvals$cors
pvals <- glory_c2.permuted.pvals$pvals
sparCCpcors_glory_c2 <- diag(0.5, nrow = dim(sparcc.matrix_glory_c2$Cor)[1], ncol = dim(sparcc.matrix_glory_c2$Cor)[1])
sparCCpcors_glory_c2[upper.tri(sparCCpcors_glory_c2, diag=FALSE)] <- cors
sparCCpcors_glory_c2 <- sparCCpcors_glory_c2 + t(sparCCpcors_glory_c2)

sparCCpval_glory_c2 <- diag(0.5, nrow = dim(sparcc.matrix_glory_c2$Cor)[1], ncol = dim(sparcc.matrix_glory_c2$Cor)[1])
sparCCpval_glory_c2[upper.tri(sparCCpval_glory_c2, diag=FALSE)] <- pvals
sparCCpval_glory_c2<- sparCCpval_glory_c2 + t(sparCCpval_glory_c2)

rownames(sparCCpcors_glory_c2) <- colnames(glory_c2)
colnames(sparCCpcors_glory_c2) <- colnames(glory_c2)
rownames(sparCCpval_glory_c2) <- colnames(glory_c2)
colnames(sparCCpval_glory_c2) <- colnames(glory_c2)

sparCCpcors_glory_c2[1:5, 1:5]
sparCCpval_glory_c2[1:5, 1:5]
```
```{r}
reordered_all_sparcc_glory_c2 <- reorder_cor_and_p(sparCCpcors_glory_c2, sparCCpval_glory_c2)
reordered_sparccCor_glory_c2 <- reordered_all_sparcc_glory_c2$r
reordered_sparccP_glory_c2<- reordered_all_sparcc_glory_c2$p

sparccCor_processed_glory_c2 <- reordered_sparccCor_glory_c2  %>% get_upper_tri() %>% reshape2::melt() %>% na.omit() %>% dplyr::rename(cor = value)
sparccP_processed_glory_c2 <- reordered_sparccP_glory_c2  %>% get_upper_tri() %>% reshape2::melt() %>% na.omit() %>% dplyr::rename(p = value)

# join the two data frames
SparccP_glory_c2 <- left_join(sparccCor_processed_glory_c2, sparccP_processed_glory_c2, by = c("Var1", "Var2")) %>%
  # # remove self correlations
  # filter(Var1 != Var2) %>% 
  # calculate the false discovery rate to adjust for multiple p values
  mutate(fdr = p.adjust(p, method = "fdr"))
# Correlation matrix with p values
sparccOkP_glory_c2 <- SparccP_glory_c2%>% filter(fdr < 0.05) 
sparccOkP_glory_c2
sparccOkP_glory_c2 %>% filter(cor>0)
sparccOkP_glory_c2 %>% filter(cor<0)
```
Plotting only significant correlations
```{r}
glory_c2_sig.net<-graph_from_data_frame(sparccOkP_glory_c2[,1:3], directed=F, vertices=NULL)
plot(glory_c2_sig.net)
```
Hub detection
```{r}
#Hub detection 
net_glory_c2<-glory_c2_sig.net
net.cn_glory_c2<-closeness(net_glory_c2) #measures centrality
net.bn_glory_c2<-betweenness(net_glory_c2) #measures centrality
net.pr_glory_c2<-page_rank(net_glory_c2)$vector 
net.hs_glory_c2<-hub_score(net_glory_c2)$vector
#Sort the species based on hubbiness score 
net.hs.sort_glory_c2<-sort(net.hs_glory_c2,decreasing=TRUE) 
#Choose the top 5 keystone species 
net.hs.top5_glory_c2<-head(net.hs.sort_glory_c2,n=5)
print(net.hs.top5_glory_c2)
```
Cluster detection
```{r}
#Get clusters 
wt_glory_c2<-walktrap.community(net_glory_c2) 
ml_glory_c2<-multilevel.community(net_glory_c2) 
#Get membership of walktrap clusters 
membership(wt_glory_c2)
wt_glory_c2membership<-membership(wt_glory_c2)
wt_glory_c2membership<-as.data.frame(as.matrix(wt_glory_c2membership))
colnames(wt_glory_c2membership)<-'WT_cluster'
#Compare clusters detected by different methods 
compare(membership(wt_glory_c2),membership(ml_glory_c2))
#Plot clusters as dendrogram 
plot_dendrogram(wt_glory_c2)
```
Modularity and articulation points
```{r}
#Calculate modularity 
modularity(net_glory_c2,membership(wt_glory_c2))
#Find articulation points 
AP_glory_c2<-articulation.points(net_glory_c2)#(Nodes so central that their removal will break the network into more components)
```
Network features
```{r}
nodes_glory_c2<-V(net_glory_c2) #vectors of nodes 
edges_glory_c2<-V(net_glory_c2) 
node.names_glory_c2<-V(net_glory_c2)$name 
num.nodes_glory_c2<-vcount(net_glory_c2) #number of nodes
print(num.nodes_glory_c2) #(number of Asvs)
num.edges_glory_c2<-ecount(net_glory_c2) #number of edges (number of lines between nodes)
print(num.edges_glory_c2)
clustering_coeff_glory_c2<-transitivity(net_glory_c2,type="global") #transensitivity
print(clustering_coeff_glory_c2)
```
Plot
```{r}
plot(wt_glory_c2,net_glory_c2) 
plot(wt_glory_c2,net_glory_c2, vertex.label.cex = 0.4, frame.color=NA, main="Younger germination cohort")
```

# Estimating heritability
## Community wide metrics
```{r}
#getting species richness and ordination first 3 axes into one metadata file
metadataphyR<-as.data.frame(sample_data(phyglory_rarefy))
gloryrichness_rarefied<-estimate_richness(phyglory_rarefy) #generating observed and shannon 
gloryrichness_rarefied<-gloryrichness_rarefied %>% mutate(evenness= Shannon/log(Observed)) #calculate evenness
metadataphyR$ObservedRich<-gloryrichness_rarefied$Observed
metadataphyR$Shannon<-gloryrichness_rarefied$Shannon
metadataphyR$evenness<-gloryrichness_rarefied$evenness

phyglory_rare_rel <- microbiome::transform(phyglory_rarefy, "compositional")
phyglory.ord.wuniR <- ordinate(phyglory_rare_rel, "PCoA", "unifrac", weighted=T) #this is how the pcoa was generated, the same one as in the main text
str(phyglory.ord.wuniR)
phyglory.ord.wuniR$values
phyglory.ord.wuniR$vectors #numbers from first 3 pcoa axes
metadataphyR$PCoA_axis1<-phyglory.ord.wuniR$vectors[,1]
metadataphyR$PCoA_axis2<-phyglory.ord.wuniR$vectors[,2]
metadataphyR$PCoA_axis3<-phyglory.ord.wuniR$vectors[,3]
metadataRH2<-as.data.frame(as(metadataphyR, "matrix")) #save phyloseq object as dataframe

str(metadataRH2)
metadataRH2$ObservedRich<-as.integer(metadataRH2$ObservedRich) #fix some of the salient categorization problems
metadataRH2$Shannon<-as.numeric(metadataRH2$Shannon)
metadataRH2$evenness<-as.numeric(metadataRH2$evenness)
metadataRH2$PCoA_axis1<-as.numeric(metadataRH2$PCoA_axis1)
metadataRH2$PCoA_axis2<-as.numeric(metadataRH2$PCoA_axis2)
metadataRH2$PCoA_axis3<-as.numeric(metadataRH2$PCoA_axis3)
metadataRH2$Germination_cohort<-as.factor(metadataRH2$Germination_cohort)
metadataRH2$Shape<-as.factor(metadataRH2$Shape)
metadataRH2$Line<-as.factor(metadataRH2$Line)
metadataRH2$Block<-as.factor(metadataRH2$Block)

hist(metadataRH2$ObservedRich)#check normality of data
hist(log(metadataRH2$ObservedRich)) 
hist(metadataRH2$Shannon)
hist(metadataRH2$PCoA_axis1)
hist(metadataRH2$PCoA_axis2)
hist(metadataRH2$PCoA_axis3)
```
Running the models
```{r}
#for our H2 estimates:
set.seed(0)
RH2_ObservedRich<-lmer(log(ObservedRich)~ Germination_cohort + Shape + (1|Line) + (1|Block), data=metadataRH2, control = lmerControl(optimizer ="Nelder_Mead"))
RH2_Shannon<-lmer(Shannon~ Germination_cohort + Shape + (1|Line) + (1|Block), data=metadataRH2, control = lmerControl(optimizer ="Nelder_Mead"))
RH2_evenness<-lmer(evenness~ Germination_cohort + Shape + (1|Line) + (1|Block), data=metadataRH2, control = lmerControl(optimizer ="Nelder_Mead"))
RH2_PCoA1<-lmer(PCoA_axis1~ Germination_cohort + Shape + (1|Line) + (1|Block), data=metadataRH2, control = lmerControl(optimizer ="Nelder_Mead"))
RH2_PCoA2<-lmer(PCoA_axis2~ Germination_cohort + Shape + (1|Line) + (1|Block), data=metadataRH2, control = lmerControl(optimizer ="Nelder_Mead"))
RH2_PCoA3<-lmer(PCoA_axis3~ Germination_cohort + Shape + (1|Line) + (1|Block), data=metadataRH2, control = lmerControl(optimizer ="Nelder_Mead"))

summary(RH2_ObservedRich) #sig germ
summary(RH2_Shannon)
summary(RH2_evenness)
summary(RH2_PCoA1) #sig germ
summary(RH2_PCoA2)#sig germ
summary(RH2_PCoA3) #sig M difference from H

Anova(RH2_ObservedRich, type=3)
Anova(RH2_evenness, type=3)
Anova(RH2_Shannon, type=3)
Anova(RH2_PCoA1, type=3)
Anova(RH2_PCoA2, type=3)
Anova(RH2_PCoA3, type=3)
```
loglik with models not containing Line (higher(and less negative values) are better)
```{r}
set.seed(0)
logLik(RH2_ObservedRich)#-116.773
logLik(lmer(log(ObservedRich)~ Germination_cohort + Shape + (1|Block), data=metadataRH2, control = lmerControl(optimizer ="Nelder_Mead")))# -116.7835
logLik(RH2_Shannon)#-114.6886
logLik(lmer(Shannon~ Germination_cohort + Shape + (1|Block), data=metadataRH2, control = lmerControl(optimizer ="Nelder_Mead")))#-114.7927
logLik(RH2_evenness) #269.2387
logLik(lmer(evenness~ Germination_cohort + Shape + (1|Block), data=metadataRH2, control = lmerControl(optimizer ="Nelder_Mead"))) #269.2387
logLik(RH2_PCoA1)# 265.7301 
logLik(lmer(PCoA_axis1~ Germination_cohort + Shape + (1|Block), data=metadataRH2, control = lmerControl(optimizer ="Nelder_Mead")))# 265.6788
logLik(RH2_PCoA2)# 329.5705
logLik(lmer(PCoA_axis2~ Germination_cohort + Shape + (1|Block), data=metadataRH2, control = lmerControl(optimizer ="Nelder_Mead")))# 329.5705
logLik(RH2_PCoA3) #422.3402
logLik(lmer(PCoA_axis3~ Germination_cohort + Shape + (1|Block), data=metadataRH2, control = lmerControl(optimizer ="Nelder_Mead"))) #422.208
#including line gives a better likelihood score except for Pcoa axis 2, which has same likelihood


With_random_effects_logLik<-c(logLik(RH2_ObservedRich),logLik(RH2_Shannon),logLik(RH2_evenness),logLik(RH2_PCoA1),logLik(RH2_PCoA2), logLik(RH2_PCoA3))
Noline_logLik<-c(logLik(lmer(log(ObservedRich)~ Germination_cohort + Shape + (1|Block), data=metadataRH2, control = lmerControl(optimizer ="Nelder_Mead"))),logLik(lmer(Shannon~ Germination_cohort + Shape + (1|Block), data=metadataRH2, control = lmerControl(optimizer ="Nelder_Mead"))),logLik(lmer(evenness~ Germination_cohort + Shape + (1|Block), data=metadataRH2, control = lmerControl(optimizer ="Nelder_Mead"))),logLik(lmer(PCoA_axis1~ Germination_cohort + Shape + (1|Block), data=metadataRH2, control = lmerControl(optimizer ="Nelder_Mead"))),logLik(lmer(PCoA_axis2~ Germination_cohort + Shape + (1|Block), data=metadataRH2, control = lmerControl(optimizer ="Nelder_Mead"))),logLik(lmer(PCoA_axis3~ Germination_cohort + Shape + (1|Block), data=metadataRH2, control = lmerControl(optimizer ="Nelder_Mead"))))
model_name<-c("Observed_richness","Shannon", "Evenness","PCoA1", "PCoA2", "PCoA3")
LogLikelihood_comparison_com<-data.frame(With_random_effects_logLik,Noline_logLik,model_name)
LogLikelihood_comparison_com<-LogLikelihood_comparison_com %>% mutate(Difference_noline_or_fullrandom= With_random_effects_logLik-Noline_logLik)
```
### Extract variance components and calculate heritability
```{r}
print(VarCorr(RH2_ObservedRich), comp=c("Variance", "Std.Dev."))#example output from VarCorr

communityphenoR<-c(RH2_ObservedRich, RH2_evenness, RH2_Shannon, RH2_PCoA1, RH2_PCoA2, RH2_PCoA3)
comphenlistR<-lapply(communityphenoR, function(x) as.data.frame(VarCorr(x))[,c("vcov", "sdcor")])
comphenframeR<- data.frame(matrix(unlist(comphenlistR), nrow=length(comphenlistR), byrow=TRUE))
colnames(comphenframeR)<-c("Line_variance","Block_variance", "Residual_variance","Line_stdev","Block_stdev", "Residual_stdev")
rownames(comphenframeR)<-c("Observed_Richness", "Evenness", "Shannon", "PCoA1", "PCoA2", "PCoA3")
comphenframeR<-comphenframeR %>% 
  mutate(Total_variance= Line_variance+Block_variance+Residual_variance,
         LinePercent=Line_variance/Total_variance,
         BlockPercent=Block_variance/Total_variance,
        ResidualPercent=Residual_variance/Total_variance )
comphenframeR
comphenPERCENTR<-gather(comphenframeR, key="Variance_component", value="Percent_variance_explained", 8:10)
comphenPERCENTR$Model<-c("Observed_Richness", "Evenness", "Shannon", "PCoA1", "PCoA2", "PCoA3", "Observed_Richness", "Evenness", "Shannon", "PCoA1", "PCoA2", "PCoA3", "Observed_Richness", "Evenness", "Shannon", "PCoA1", "PCoA2", "PCoA3") 
head(comphenframeR)
comphenPERCENTR
```
figure
```{r}
comphenPERCENTR$Model<-factor(comphenPERCENTR$Model, levels=c( "PCoA3", "Shannon","PCoA1", "Observed_Richness", "Evenness", "PCoA2"))
communityphenotypes<-comphenPERCENTR %>% 
  filter(!Variance_component %in% c('ResidualPercent', 'GeneticPercent')) %>% 
  ggplot(aes(x = Model, y = Percent_variance_explained, fill=Variance_component)) +
 geom_bar(stat ="identity",width=0.5) +
 coord_flip() +
 labs(x = "Community phenotype", y = "Percent variance explained")+scale_fill_manual(name="Variance component", labels=c("Spatial Block", "Plant Line"),  values = c("#FF6699", "#3399FF", "#E69F00")) +theme_cowplot() +theme(legend.position = "none")+scale_x_discrete(labels=c("PCoA axis 3","Shannon diversity","PCoA axis 1","Observed richness", "Evenness","PCoA axis 2"))
communityphenotypes
```

### code for figure comparing magnitude of line vs shape
```{r}
#for figure purposes, comparing the magnitude of effect of line vs leaf shape:
set.seed(0)
fig_ObservedRich<-lmer(log(ObservedRich)~ Germination_cohort + (1|Shape) + (1|Line) + (1|Block), data=metadataRH2, control = lmerControl(optimizer ="Nelder_Mead"))
fig_Shannon<-lmer(Shannon~ Germination_cohort + (1|Shape) + (1|Line) + (1|Block), data=metadataRH2, control = lmerControl(optimizer ="Nelder_Mead"))
fig_evenness<-lmer(evenness~ Germination_cohort + (1|Shape) + (1|Line) + (1|Block), data=metadataRH2, control = lmerControl(optimizer ="Nelder_Mead"))
fig_PCoA1<-lmer(PCoA_axis1~ Germination_cohort + (1|Shape) + (1|Line) + (1|Block), data=metadataRH2, control = lmerControl(optimizer ="Nelder_Mead"))
fig_PCoA2<-lmer(PCoA_axis2~ Germination_cohort + (1|Shape) + (1|Line) + (1|Block), data=metadataRH2, control = lmerControl(optimizer ="Nelder_Mead"))
fig_PCoA3<-lmer(PCoA_axis3~ Germination_cohort + (1|Shape) + (1|Line) + (1|Block), data=metadataRH2, control = lmerControl(optimizer ="Nelder_Mead"))

summary(fig_ObservedRich)
summary(fig_Shannon)
summary(fig_PCoA1) #
summary(fig_PCoA2)
summary(fig_PCoA3)

communityphenofig<-c(fig_ObservedRich, fig_evenness, fig_Shannon, fig_PCoA1, fig_PCoA2, fig_PCoA3)
comphenlistfig<-lapply(communityphenofig, function(x) as.data.frame(VarCorr(x))[,c("vcov", "sdcor")])
comphenframefig<- data.frame(matrix(unlist(comphenlistfig), nrow=length(comphenlistfig), byrow=TRUE))
colnames(comphenframefig)<-c("Line_variance","Block_variance","Shape_variance", "Residual_variance","Line_stdev","Block_stdev","Shape_stdev", "Residual_stdev")
rownames(comphenframefig)<-c("Observed_Richness", "Evenness", "Shannon", "PCoA1", "PCoA2", "PCoA3")
comphenframefig<-comphenframefig %>% 
  mutate(Total_variance= Line_variance+Block_variance+Shape_variance+Residual_variance,
         LinePercent=Line_variance/Total_variance,
         BlockPercent=Block_variance/Total_variance,
         ShapePercent=Shape_variance/Total_variance,
         GeneticPercent=(Line_variance+Shape_variance)/Total_variance,
        ResidualPercent=Residual_variance/Total_variance )
comphenframefig
comphenPERCENTfig<-gather(comphenframefig, key="Variance_component", value="Percent_variance_explained", 10:14)
comphenPERCENTfig$Model<-c("Observed_Richness", "Evenness", "Shannon", "PCoA1", "PCoA2", "PCoA3", "Observed_Richness", "Evenness", "Shannon", "PCoA1", "PCoA2", "PCoA3", "Observed_Richness", "Evenness", "Shannon", "PCoA1", "PCoA2", "PCoA3", "Observed_Richness", "Evenness", "Shannon", "PCoA1", "PCoA2", "PCoA3",  "Observed_Richness", "Evenness", "Shannon", "PCoA1", "PCoA2", "PCoA3")
head(comphenframefig)
```
magnitude figure
```{r}
comphenPERCENTfig$Model<-factor(comphenPERCENTfig$Model, levels=c( "PCoA3", "Shannon","PCoA1", "Observed_Richness", "Evenness", "PCoA2"))
comphenPERCENTfig$Variance_component <- factor(comphenPERCENTfig$Variance_component, levels = c("BlockPercent","ShapePercent", "LinePercent", "ResidualPercent", "GeneticPercent"))
communityphenotypesfig<-comphenPERCENTfig %>% 
  filter(!Variance_component %in% c('ResidualPercent', 'GeneticPercent')) %>% 
  ggplot(aes(x = Model, y = Percent_variance_explained, fill=Variance_component)) +
 geom_bar(stat ="identity",width=0.5) +
 coord_flip() +
 labs(x = "Community phenotype", y = "Percent variance explained")+scale_fill_manual(name="Variance component", labels=c("Spatial block", "Leaf shape genotype","Plant line"),  values = c("#FF6699","#E69F00", "#3399FF","#99CC00")) +theme_cowplot()  +theme(legend.position = "none")+scale_x_discrete(labels=c("PCoA axis 3","Shannon diversity","PCoA axis 1","Observed richness", "Evenness", "PCoA axis 2"))
communityphenotypesfig
```

## Genera CLR transformed
```{r}
metadataH2<-as.data.frame(sample_data(phyglory))
#write.csv(metadataH2, "metadataH2.csv")
metadataH2<-read.csv('metadataH2.csv',row.names = 1, header = TRUE)#to get it as a dataframe, not neccessary for the model though, it can be left in phyloseq-ish format
metadataH2$Germination_cohort<-as.factor(metadataH2$Germination_cohort)
metadataH2$Shape<-as.factor(metadataH2$Shape)
metadataH2$Line<-as.factor(metadataH2$Line)
metadataH2$Block<-as.factor(metadataH2$Block)
metadataH2 #this metadata has corresponding rows to the feature table (100, 101, 102... 10 etc.)

str(metadataH2)
metadataH2
```
### Models and figure
```{r}
featuretableCLRgen
my_lmersCLR <- lapply(1:15, function(x) lmer(featuretableCLRgen[,x]~ metadataH2$Germination_cohort +metadataH2$Shape + (1|metadataH2$Line) + (1|metadataH2$Block), control = lmerControl(optimizer ="Nelder_Mead"))) 

# call the summaries of the list of regressions and make a dataframe:
summariesCLR <- lapply(my_lmersCLR, summary)
head(summariesCLR)
# list of only main coefficents with estimates and with p values for ipomoeajuice:
summaries_subset_listCLR<- lapply(summariesCLR, function(x) x$coefficients[, c(1,5)])
head(summaries_subset_listCLR)
summaries_subset_frameCLR <- data.frame(matrix(unlist(summaries_subset_listCLR), nrow=length(summaries_subset_listCLR), byrow=TRUE))
colnames(summaries_subset_frameCLR) <- c("Intercept_estimate","Germ_cohort_estimate","ShapeL_estimate","ShapeM_estimate","Intercept_pvalue", "Germ_cohort_pvalue", "ShapeL_pvalue", "ShapeM_pvalue")
head(summaries_subset_frameCLR)

#VarCorr to extract random effects 
summaries_subset_var_listCLR<-lapply(my_lmersCLR, function(x) as.data.frame(VarCorr(x))[,c("vcov", "sdcor")])
summaries_subset_var_frameCLR<- data.frame(matrix(unlist(summaries_subset_var_listCLR), nrow=length(summaries_subset_var_listCLR), byrow=TRUE))
head(summaries_subset_var_frameCLR)
colnames(summaries_subset_var_frameCLR)<-c("Line_variance","Block_variance","Residual_variance","Line_stdev","Block_stdev","Residual_stdev")
head(summaries_subset_var_frameCLR)

#Running type 3 anovas on the models to get F test statistics for shape
type3_lmersCLR<-lapply(1:15, function(x) Anova((lmer(featuretableCLRgen[,x]~ metadataH2$Germination_cohort +metadataH2$Shape + (1|metadataH2$Line) + (1|metadataH2$Block), control = lmerControl(optimizer ="Nelder_Mead"))), type=3))
type3_lmersCLR
type3_lmers_frameCLR<-data.frame(matrix(unlist(type3_lmersCLR), nrow=length(type3_lmersCLR), byrow=TRUE))
type3_lmers_frameCLR
colnames(type3_lmers_frameCLR) <- c("Intercept_Chisq","Germ_cohort_Chisq","Shape_Chisq","Intercept_df", "Germ_cohort_df","Shape_df", "Intercept_pvalue", "Germ_cohort_pvalue", "Shape_pvalue" )
Germ_cohort_anovap<-type3_lmers_frameCLR[,8]
Leafshape_anovap<-type3_lmers_frameCLR[,9]

summaries_outputCLR<-cbind(summaries_subset_frameCLR, summaries_subset_var_frameCLR) #combine fixed and random outputs
summaries_output_fullCLR<-cbind(summaries_outputCLR, Germ_cohort_anovap)
summaries_output_fullCLR<-cbind(summaries_output_fullCLR, Leafshape_anovap)
genus<-colnames(featuretableCLRgen)
model_number<-seq(1:500)
summaries_output_fullCLR$Genus<-genus[1:15] #connect model outputs to the ASV it was run on
summaries_output_fullCLR$Model_number<-model_number[1:15] #keep track of what model output we're looking at. Doubles as commonness of ASV
summaries_output_fullCLR
ipomoeajuiceCLR<-summaries_output_fullCLR

ipomoeajuiceCLR<-ipomoeajuiceCLR %>%  
  mutate(Total_variance= Line_variance+Block_variance+Residual_variance,
         LinePercent=Line_variance/Total_variance,
         BlockPercent=Block_variance/Total_variance,
        ResidualPercent=Residual_variance/Total_variance )
ipomoeajuiceCLR
ipomoeaPERCENTCLR<-gather(ipomoeajuiceCLR, key="Variance_component", value="Percent_variance_explained", 20:22)
ipomoeaPERCENTCLR<-ipomoeaPERCENTCLR %>% 
  mutate(Line_percent=(Line_variance)/Total_variance)

genera<-ipomoeaPERCENTCLR %>% 
   filter(!Variance_component %in% c('ResidualPercent', 'Line_percent')) %>% 
  ggplot(aes(x=reorder(Genus, -Line_percent), y=Percent_variance_explained,fill=Variance_component))+
 geom_bar(stat ="identity",width=.75) +
 coord_flip() +
 labs(x = "Genus", y = "Percent variance explained")+scale_fill_manual(name="Variance component", labels=c("Spatial block", "Plant line"),  values = c("#FF6699", "#3399FF", "#E69F00")) +theme_cowplot() + theme(axis.text.y = element_text(face="italic"))
genera
```
### log likelihood for H2 models
```{r}
my_lmers_likelihood <-lapply(1:15, function(x) logLik( lmer(featuretableCLRgen[,x]~ metadataH2$Germination_cohort +metadataH2$Shape + (1|metadataH2$Line) + (1|metadataH2$Block), control = lmerControl(optimizer ="Nelder_Mead"))) )

my_lmers_likelihood_noline <- lapply(1:15, function(x) logLik( lmer(featuretableCLRgen[,x]~ metadataH2$Germination_cohort +metadataH2$Shape + (1|metadataH2$Block), control = lmerControl(optimizer ="Nelder_Mead"))) )

my_lmers_likelihood_frame<-data.frame(matrix(unlist(my_lmers_likelihood), nrow=length(my_lmers_likelihood), byrow=TRUE))
my_lmers_likelihood_noline_frame<-data.frame(matrix(unlist(my_lmers_likelihood_noline), nrow=length(my_lmers_likelihood_noline), byrow=TRUE))

colnames(my_lmers_likelihood_frame) <- c("With_random_effects_logLik")
colnames(my_lmers_likelihood_noline_frame) <- c("Noline_logLik")

LogLikelihood_comparison<-cbind(my_lmers_likelihood_frame, my_lmers_likelihood_noline_frame)
LogLikelihood_comparison$Model_number<-model_number[1:15]
LogLikelihood_comparison$model_name<-genus[1:15]

LogLikelihood_comparison<-LogLikelihood_comparison %>% mutate(Difference_noline_or_fullrandom= With_random_effects_logLik-Noline_logLik)#if 'noline' is more more negative/worse, then the value will be positive, positive value means line improved fit
#View(LogLikelihood_comparison)

LogLikelihood_comparison_all<-rbind(LogLikelihood_comparison,LogLikelihood_comparison_com)
LogLikelihood_comparison_all$Model_number<-model_number[1:21]
 
LogLikelihood_comparison %>% 
  filter(Difference_noline_or_fullrandom>0)#positive value means line improved fit
LogLikelihood_comparison %>% 
  filter(Difference_noline_or_fullrandom<0) #very slight worse fit, but the same to 4 decimals
LogLikelihood_comparison %>% 
  filter(Difference_noline_or_fullrandom==0) # exact same fit
```

### magnitude line vs leaf shape models and figure
```{r}
featuretableCLRgen
my_lmersfig <- lapply(1:15, function(x) lmer(featuretableCLRgen[,x]~ metadataH2$Germination_cohort +(1|metadataH2$Shape) + (1|metadataH2$Line) + (1|metadataH2$Block), control = lmerControl(optimizer ="Nelder_Mead"))) 

# call the summaries of the list of regressions and make a dataframe:
summariesfig <- lapply(my_lmersfig, summary)
summariesfig
# list of only main coefficents with estimates and with p values for ipomoeajuice:
summaries_subset_listfig<- lapply(summariesfig, function(x) x$coefficients[, c(1,5)])
head(summaries_subset_listfig)
summaries_subset_framefig <- data.frame(matrix(unlist(summaries_subset_listfig), nrow=length(summaries_subset_listfig), byrow=TRUE))
colnames(summaries_subset_framefig) <- c("Intercept_estimate","Germ_cohort_estimate","Intercept_pvalue", "Germ_cohort_pvalue" )
head(summaries_subset_framefig)

#VarCorr to extract random effects 
summaries_subset_var_listfig<-lapply(my_lmersfig, function(x) as.data.frame(VarCorr(x))[,c("vcov", "sdcor")])
summaries_subset_var_framefig<- data.frame(matrix(unlist(summaries_subset_var_listfig), nrow=length(summaries_subset_var_listfig), byrow=TRUE))
head(summaries_subset_var_framefig)
colnames(summaries_subset_var_framefig)<-c("Line_variance","Block_variance","Shape_variance", "Residual_variance","Line_stdev","Block_stdev","Shape_stdev", "Residual_stdev")
head(summaries_subset_var_framefig)

#Running type 3 anovas on the models to get F test statistics for shape
type3_lmersfig<-lapply(1:15, function(x) Anova((lmer(featuretableCLRgen[,x]~ metadataH2$Germination_cohort +(1|metadataH2$Shape) + (1|metadataH2$Line) + (1|metadataH2$Block), control = lmerControl(optimizer ="Nelder_Mead"))), type=3))
type3_lmers_framefig<-data.frame(matrix(unlist(type3_lmersfig), nrow=length(type3_lmersfig), byrow=TRUE))
type3_lmers_framefig
colnames(type3_lmers_framefig) <- c("Intercept_Chisq","Germ_cohort_Chisq","Intercept_df", "Germ_cohort_df", "Intercept_pvalue", "Germ_cohort_pvalue" )
Germ_cohort_anovap<-type3_lmers_framefig[,6]

summaries_outputfig<-cbind(summaries_subset_framefig, summaries_subset_var_framefig) #combine fixed and random outputs
summaries_output_fullfig<-cbind(summaries_outputfig, Germ_cohort_anovap)
genus<-colnames(featuretableCLRgen)
model_number<-seq(1:500)
summaries_output_fullfig$Genus<-genus[1:15] #connect model outputs to the ASV it was run on
summaries_output_fullfig$Model_number<-model_number[1:15] #keep track of what model output we're looking at. Doubles as commonness of ASV
summaries_output_fullfig
ipomoeajuicefig<-summaries_output_fullfig

ipomoeajuicefig<-ipomoeajuicefig %>%  
  mutate(Total_variance= Line_variance+Block_variance+Shape_variance+Residual_variance,
         LinePercent=Line_variance/Total_variance,
         BlockPercent=Block_variance/Total_variance,
         ShapePercent=Shape_variance/Total_variance,
         GeneticPercent=(Line_variance+Shape_variance)/Total_variance,
        ResidualPercent=Residual_variance/Total_variance )
ipomoeajuicefig
ipomoeaPERCENTfig<-gather(ipomoeajuicefig, key="Variance_component", value="Percent_variance_explained", 17:21)
ipomoeaPERCENTfig<-ipomoeaPERCENTfig %>% 
  mutate(Genetic_percent=(Line_variance+Shape_variance)/Total_variance)
ipomoeaPERCENTfig<-ipomoeaPERCENTfig %>% 
  mutate(Line_percent=(Line_variance)/Total_variance)
ipomoeaPERCENTfig
ipomoeaPERCENTfig$Variance_component <- factor(ipomoeaPERCENTfig$Variance_component, levels = c("BlockPercent","ShapePercent", "LinePercent", "ResidualPercent", "GeneticPercent"))
ipomoeaPERCENTfig$Variance_component

generafig<-ipomoeaPERCENTfig %>% 
   filter(!Variance_component %in% c('ResidualPercent', 'GeneticPercent')) %>% 
  ggplot(aes(x=reorder(Genus, -ipomoeaPERCENTCLR$Line_percent), y=Percent_variance_explained,fill=Variance_component))+
 geom_bar(stat ="identity",width=.75) +
 coord_flip() +
 labs(x = "Genus", y = "Percent variance explained")+scale_fill_manual(name="Variance component", labels=c("Spatial block", "Leaf shape genotype","Plant line"),  values = c("#FF6699","#E69F00", "#3399FF","#99CC00")) +theme_cowplot() + theme(axis.text.y = element_text(face="italic")) 
generafig
```

# Figures
ancom figure
```{r}
ggsave2("ancomglory_shape_plot.png", ancomglory_shape_plot, dpi=600, width=7, height=5)
ancomglory_shape_plot
```

composition barplot
```{r}
multi_leaf <- (leaf_germ1 + leaf_germ2 ) + 
  plot_layout(nrow=1)+
  plot_annotation(tag_levels = list(c("A", "B"))) #add figure labels
multi_leaf #save 5 by 15
```

succession figure
```{r}
#possible succession figure
a.divglory+(b.div.wuniglory_rareG/b.div.wuniglory_rare)+plot_annotation(tag_levels="A") #save 6 by 10
```

heritability figure
```{r}
plot_grid(communityphenotypes,genera, labels="AUTO" , rel_widths = 2:3) #save 5 by 10 inches
```

magnitude of line vs shape figure
```{r}
plot_grid(communityphenotypesfig,generafig, labels="AUTO" , rel_widths = 2:3) #save 5 by 10 inches
```
all heritability figures?
```{r}
communityphenotypes<-comphenPERCENTR %>% 
  filter(!Variance_component %in% c('ResidualPercent', 'GeneticPercent')) %>% 
  ggplot(aes(x = Model, y = Percent_variance_explained, fill=Variance_component)) +
 geom_bar(stat ="identity",width=0.5) +
 coord_flip() +
 labs(x = "Community phenotype", y = "Percent variance explained")+scale_fill_manual(name="Variance component", labels=c("Spatial Block", "Plant Line"),  values = c("#FF6699", "#3399FF", "#E69F00")) +theme_cowplot() +theme(legend.position = "none")+scale_x_discrete(labels=c("PCoA axis 3","Shannon diversity","PCoA axis 1","Observed richness", "Evenness", "PCoA axis 2"))+ylim(0,0.15)
communityphenotypes
communityphenotypesfig<-comphenPERCENTfig %>% 
  filter(!Variance_component %in% c('ResidualPercent', 'GeneticPercent')) %>% 
  ggplot(aes(x = Model, y = Percent_variance_explained, fill=Variance_component)) +
 geom_bar(stat ="identity",width=0.5) +
 coord_flip() +
 labs(x = "Community phenotype", y = "Percent variance explained")+scale_fill_manual(name="Variance component", labels=c("Spatial block", "Leaf shape genotype","Plant line"),  values = c("#FF6699","#E69F00", "#3399FF","#99CC00")) +theme_cowplot()  +theme(legend.position = "none")+scale_x_discrete(labels=c("PCoA axis 3","Shannon diversity","PCoA axis 1","Observed richness", "Evenness", "PCoA axis 2"))+ylim(0,0.15)
communityphenotypesfig
genera<-ipomoeaPERCENTCLR %>% 
   filter(!Variance_component %in% c('ResidualPercent', 'Line_percent')) %>% 
  ggplot(aes(x=reorder(Genus, -Line_percent), y=Percent_variance_explained,fill=Variance_component))+
 geom_bar(stat ="identity",width=.75) +
 coord_flip() +
 labs(x = "Genus", y = "Percent variance explained")+scale_fill_manual(name="Variance component", labels=c("Spatial block", "Plant line"),  values = c("#FF6699", "#3399FF", "#E69F00")) +theme_cowplot() +ylim(0,0.13) + theme(axis.text.y = element_text(face="italic"))
genera
generafig<-ipomoeaPERCENTfig %>% 
   filter(!Variance_component %in% c('ResidualPercent', 'GeneticPercent')) %>% 
  ggplot(aes(x=reorder(Genus, -ipomoeaPERCENTCLR$Line_percent), y=Percent_variance_explained,fill=Variance_component))+
 geom_bar(stat ="identity",width=.75) +
 coord_flip() +
 labs(x = "Genus", y = "Percent variance explained")+scale_fill_manual(name="Variance component", labels=c("Spatial block", "Leaf shape genotype","Plant line"),  values = c("#FF6699","#E69F00", "#3399FF","#99CC00")) +theme_cowplot()  +ylim(0,0.13) + theme(axis.text.y = element_text(face="italic"))
generafig

plot_grid((communityphenotypes),genera,(communityphenotypesfig),generafig, labels="AUTO" , rel_widths = 2:3) #saved 10 by 10
```

pcoa figure
```{r}
plot_grid(b.div.wuniglory_rareG, b.div.wuniglory_rare,b.div.wuniglory_methylo,b.div.wuniglory_sphingo,b.div.wuniglory_art, b.div.wuniglory_dein, ncol=2, labels="AUTO") #save 8 by 13 for only taxa. With other plots save 13x13
```

# Supplemental figures
leaf morphology 
```{r}
plot_grid(area,circularity, labels="AUTO", rel_widths = 3:2) #save 5 by 10 inches
```

relative abundances and regressions
```{r}
a<-ggplot(aes(x=sample_data(phyglory_rel_rare)$Methylobacterium_relabundance, y=sample_data(phyglory_rel_rare)$Sphingomonas_relabundance), data=metadataphyR)+geom_point()+geom_smooth(method='lm')+mdthemes::md_theme_classic() +labs(x="*Methylobacterium* relative abundance", y="*Sphingomonas* relative abundance")
b<-ggplot(aes(x=PCoA_axis1, y=sample_data(phyglory_rel_rare)$Methylobacterium_relabundance), data=metadataphyR)+geom_point()+geom_smooth(method='lm')+mdthemes::md_theme_classic()+labs(title="*Methylobacterium*",x="Axis 1 [53.2%]",y="Relative abundance")
c<-ggplot(aes(x=PCoA_axis2, y=sample_data(phyglory_rel_rare)$Sphingomonas_relabundance), data=metadataphyR)+geom_point()+geom_smooth(method='lm')+mdthemes::md_theme_classic()+labs(title="*Sphingomonas*",x="Axis 2 [23.2%]",y="Relative abundance")

a+(b/c) +plot_annotation(tag_levels = list(c("A","B", "C"))) #saved 5 by 8 landscape
```
Permutations of p values
```{r}
plot(hist1, xlab=("Germination cohort p value"), main=("PERMANOVA"))
plot(hist2, xlab=("Germination cohort p value"), main=("PERMADISP"))
```

Log likelihoods
```{r}
LogLikelihood_comparison_all %>% 
  ggplot(aes(x=reorder(model_name, Model_number), y=Difference_noline_or_fullrandom, color=ifelse(Difference_noline_or_fullrandom >0, "Improves model fit", "Same or worse model fit")))+
  geom_point()+  scale_color_manual(name="Effect of including line", values= c("red", "black"))+ xlab('Model') +ylab('Log likelihood difference') +theme_bw()+theme(axis.text.x = element_text(face="italic",size=10, angle = 45, hjust = 1))  #saved landscape 4x6inches
```

